cpp/ql/src/Likely Bugs/Arithmetic/SignedOverflowCheck.qhelp

```diff
@@ -0,0 +1,70 @@
+<!DOCTYPE qhelp PUBLIC
+  "-//Semmle//qhelp//EN"
+  "qhelp.dtd">
+<qhelp>
+<overview>
+<p>
+When checking for integer overflow, one often writes tests like
+<code>a + b &lt; a</code>.  This works fine if <code>a</code> or
+<code>b</code> are unsigned integers, since any overflow in the addition
+will cause the value to simply "wrap around".  However, using
+<i>signed</i> integers is problematic because signed overflows have undefined
+behavior according to the C and C++ standards.  If the addition overflows
+and has an undefined result, the comparison will likewise be undefined;
+it may produce an unintended result, or may be deleted entirely by an
+optimizing compiler.
+</p>
+</overview>
+<recommendation>
+<p>
+When checking for overflow by adding two values, first make sure that <code>a</code>
+or <code>b</code> are (converted into) unsigned values, unless it is
+certain that the signed addition cannot overflow.
```

Comment:
> My point about recommendations is that it's not good enough to give examples of some overflow checks that are bad and some overflow checks that are good.

Why not?  What kind of assumptions are we making about the competence of the developer?

> In particular, just casting the variables to unsigned is not a fix because it changes the overflow check so it's now doing something else. For `int a`, the expression `a + 1 < a` is intended to be false when `a == 0x7fffffff` and true otherwise. We can't recommend changing that to `(unsigned)a + 1 < a` since that's instead false when `a == 0xffffffff` and true otherwise.

The example is contrived as it involves a constant addend, but even then it can be used to illustrate my point.  Given `s + 1 < s` using signed arithmetic, how would one express it using unsigned values?  Assuming that the developer "reasonably" expects that `s + 1` will never be undefined, the answer seems to be `u !== 0x7fffffff`.  What if we instead have `s + 10 < s`? We then would have `u < 0x7ffffff6 || u > 0x7fffffff`.  Is this the recommendation we should offer developers?  For an arbitrary `s + d < s`, we would have something like `u < INT_MAX - d + 1 || u > 0x7fffffff`.

So I guess my point here is that, in the general case, there does not exist a clean signed-to-unsigned transformation that preserves the semantics of the former.  When we tell the developer that their expression has undefined behavior, we are therefore asking them to _rethink_ their design.  For example, `s + 10 < s`  could be rewritten as `s >= INT_MAX - 9` (or, more likely, `s  + 10 > s` would become `s < INT_MAX - 9`).

Reply:
You've pretty much written the advice we should give: tell the developer to replace `s + 10 < s` with `s > INT_MAX - 10`. That seems clean and simple to me. It's also what @geoffw0 suggested in the Slack link I put in my earlier review. Notice that there is no unsigned arithmetic going on here at all -- only signed but well-defined arithmetic.

Cases for non-constant deltas will be more difficult. I googled it and found http://www.cs.utah.edu/~regehr/papers/overflow12.pdf, which would be an excellent addition to the qhelp references. It spells out how to check for `signed int` addition overflow in III.B.1. Instead of applying that advice, I'd encourage developers to redesign their data types to be unsigned; the problem is just that changing a declaration from signed to unsigned may not be possible without breaking an API.

Reply:
Thank you for the paper reference; I'll be sure to include it in Qhelp.

I'm thinking about the Recommendation section, and perhaps the best way of constructing it is to include a table mapping out possible transformations.  What do you think?  It seems that we have gotten bogged down in "i before e except after c" discussions, and a table would greatly simplify things.

Reply:
A table sounds good. If the qhelp file format doesn't actually have support for table mark-up, then at least a bullet list might work.

Reply:
Tables seem to work